tools/common/Driver.cs

```diff
@@ -166,8 +168,51 @@ static void AddSharedOptions (Application app, Mono.Options.OptionSet options)
 			}
 		}
 
+		static int verbose = GetDefaultVerbosity ();
 		public static int Verbosity {
 			get { return verbose; }
+			set {
+				verbose = value;
+				ErrorHelper.Verbosity = Verbosity;
+			}
+		}
+
+		static int GetDefaultVerbosity ()
+		{
+			var v = 0;
+			var fn = Path.Combine (Environment.GetFolderPath (Environment.SpecialFolder.Personal), $".{NAME}-verbosity");
+			if (File.Exists (fn)) {
+				v = (int) new FileInfo (fn).Length;
+				if (v == 0)
+					v = 4; // this is the magic verbosity level we give everybody.
+			}
+			return v;
+		}
+
+		public static void Log (string value)
+		{
+			Log (0, value);
+		}
+
+		public static void Log (string format, params object [] args)
+		{
+			Log (0, format, args);
+		}
+
+		public static void Log (int min_verbosity, string value)
+		{
+			if (min_verbosity > Verbosity)
+				return;
+
+			Console.WriteLine (value);
+		}
+
+		public static void Log (int min_verbosity, string format, params object [] args)
+		{
+			if (min_verbosity > Verbosity)
+				return;
+
+			Console.WriteLine (format, args);
```

Comment:
This is incorrect and already a bug https://github.com/xamarin/xamarin-macios/issues/7904

We can end up with a `format` that has `{x}` inside it and no `args` - since a file could be named `{x}.cs` (even if the issue above is related to a different issue).

My local fix for this is
```diff
diff --git a/tools/mmp/driver.cs b/tools/mmp/driver.cs
index 59fcf750c..84d5a6ab6 100644
--- a/tools/mmp/driver.cs
+++ b/tools/mmp/driver.cs
@@ -1041,7 +1041,10 @@ namespace Xamarin.Bundler {
                        if (min_verbosity > verbose)
                                return;
 
-                       Console.WriteLine (format, args);
+                       if (args.Length > 0)
+                               Console.WriteLine (format, args);
+                       else
+                               Console.WriteLine (format);
                }
 
                static string GenerateMain ()
```

but since you're moving/unifying the code it might be better to fix it inside this PR.

Reply:
Fix (and test) in https://github.com/xamarin/xamarin-macios/pull/7941